File: src/main/java/com/fengsheng/network/Network.java
package com.fengsheng.network;

import com.fengsheng.Config;
import io.netty.bootstrap.ServerBootstrap;
import io.netty.channel.ChannelFuture;
import io.netty.channel.EventLoopGroup;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.nio.NioServerSocketChannel;

import java.util.concurrent.CountDownLatch;

public final class Network {
    public static void init() {
        new Thread(Network::initGameNetwork).start();
        new Thread(Network::initGmNetwork).start();
        try {
            cd.await();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    private static final CountDownLatch cd = new CountDownLatch(2);

    private static void initGameNetwork() {
        EventLoopGroup bossGroup = new NioEventLoopGroup();
        EventLoopGroup workerGroup = new NioEventLoopGroup();
        try {
            ServerBootstrap bootstrap = new ServerBootstrap();

            bootstrap.group(bossGroup, workerGroup)
                    .channel(NioServerSocketChannel.class)
                    .childHandler(new ProtoServerInitializer());

            ChannelFuture future = bootstrap.bind(Config.ListenPort);
            cd.countDown();
            future.channel().closeFuture().sync();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        } finally {
            bossGroup.shutdownGracefully();
            workerGroup.shutdownGracefully();
        }
    }

    private static void initGmNetwork() {
        if (!Config.IsGmEnable) {
            cd.countDown();
            return;
        }
        EventLoopGroup bossGroup = new NioEventLoopGroup();
        EventLoopGroup workerGroup = new NioEventLoopGroup();
        try {
            ServerBootstrap bootstrap = new ServerBootstrap();

            bootstrap.group(bossGroup, workerGroup)
                    .channel(NioServerSocketChannel.class)
                    .childHandler(new HttpServerInitializer());

            ChannelFuture future = bootstrap.bind(Config.GmListenPort);
            cd.countDown();
            future.channel().closeFuture().sync();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        } finally {
            bossGroup.shutdownGracefully();
            workerGroup.shutdownGracefully();
        }
    }
}

File: src/main/kotlin/network/Network.kt
package com.fengsheng.network

import com.fengsheng.Config
import io.netty.bootstrap.ServerBootstrap
import io.netty.channel.EventLoopGroup
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.nio.NioServerSocketChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.joinAll
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

object Network {
    fun init() {
        runBlocking {
            listOf(
//                launch(Dispatchers.IO) { initGameNetwork() },
                launch(Dispatchers.IO) { initGameWebSocketNetwork() },
                launch(Dispatchers.IO) { initGmNetwork() },
            ).joinAll()
        }
    }

//    private fun initGameNetwork() {
//        val bossGroup: EventLoopGroup = NioEventLoopGroup()
//        val workerGroup: EventLoopGroup = NioEventLoopGroup()
//        try {
//            val bootstrap = ServerBootstrap()
//            bootstrap.group(bossGroup, workerGroup)
//                .channel(NioServerSocketChannel::class.java)
//                .childHandler(ProtoServerInitializer())
//            val future = bootstrap.bind(Config.ListenPort)
//            future.addListener { channelFuture ->
//                channelFuture.isSuccess || throw channelFuture.cause()
//            }
//            future.channel().closeFuture().sync()
//        } finally {
//            bossGroup.shutdownGracefully()
//            workerGroup.shutdownGracefully()
//        }
//    }

    private fun initGameWebSocketNetwork() {
        val bossGroup: EventLoopGroup = NioEventLoopGroup()
        val workerGroup: EventLoopGroup = NioEventLoopGroup()
        try {
            val bootstrap = ServerBootstrap()
            bootstrap.group(bossGroup, workerGroup)
                .channel(NioServerSocketChannel::class.java)
                .childHandler(WebSocketServerInitializer())
            val future = bootstrap.bind(Config.ListenWebSocketPort)
            future.addListener { channelFuture ->
                channelFuture.isSuccess || throw channelFuture.cause()
            }
            future.channel().closeFuture().sync()
        } finally {
            bossGroup.shutdownGracefully()
            workerGroup.shutdownGracefully()
        }
    }

    private fun initGmNetwork() {
        val bossGroup: EventLoopGroup = NioEventLoopGroup()
        val workerGroup: EventLoopGroup = NioEventLoopGroup()
        try {
            val bootstrap = ServerBootstrap()
            bootstrap.group(bossGroup, workerGroup)
                .channel(NioServerSocketChannel::class.java)
                .childHandler(HttpServerInitializer())
            val future = bootstrap.bind(Config.GmListenPort)
            future.addListener { channelFuture ->
                channelFuture.isSuccess || throw channelFuture.cause()
            }
            future.channel().closeFuture().sync()
        } finally {
            bossGroup.shutdownGracefully()
            workerGroup.shutdownGracefully()
        }
    }
}
